### ISSUE TITLE
DLL's built with MSVC toolchain don't work

### ISSUE THREAD

**Issue by mcnameej** [libressl-portable/portable#223]
_TL;dr version:_ LibreSSL DLL's built with MSVC don't export any symbols, and are therefore useless.  A one line addition to CMakeLists.txt can solve 95% of the problem, but a 100% fix requires LibreSSL code (header file) changes.  (Note that anything less than 100% is not a fix.)
---

I'm using LibreSSL 2.4.2 via the released tarball (not a git checkout) and CMake version 3.6.1.

DLL's built using the MSVC toolchain don't export any symbols.  The outcome is the same regardless of the CMake generator / build system (e.g. it fails with both Visual Studio and Ninja).

As you probably know, MSVC requires all symbols exported from a DLL to be explicitly declared, either in the source code with __declspec, or in a DEF file passed to the linker.  The LibreSSL code has no __declspec annotations and no DEF file, so nothing is exported.

You may not see the problem at first, because there is no error during the build.  The DLL's just don't work.  The openssl.exe that gets built works fine, because it ends up being linked against the static libraries, rather than the DLL's.  Building the DLL's should generate a matching import library, which would overwrite the static library (since they both have the same name, which is a whole separate problem).  However, since the DLL's don't export anything, no import library is created, and the static libraries remain intact.

CMake 3.4 and above will create a .DEF file automatically if the **WINDOWS_EXPORT_ALL_SYMBOLS** property is set on the target.  For example:

``` cmake
set_target_properties(crypto-shared PROPERTIES WINDOWS_EXPORT_ALL_SYMBOLS true)
```

It finds all the public functions in all the object files, and creates a DEF file to export them.  This is a very "un-Windows" thing to do, and results in "over sharing" (it exports functions that are only intended for use within the DLL), but may be OK for LibreSSL.

Adding that to CMakeLists.txt in _crypto_, _ssl_, and _tls_ solves about 95% of the problem.  I got the DLL's to build with all the functions exported, but was unable to link openssl.exe.  I was missing a few symbols that turned out to be data.  Solving that will be more difficult.

CMake EXPORT_ALL only exports **functions**, not **data** (e.g. global variables).  Exporting data from a Windows DLL is complicated, and CMake is right not to attempt it.  You can export data symbols in a DEF file, but it probably won't end well.

IMHO, the only sane way to handle exported data on Windows is to annotate the declarations in the header file.  Use __declspec(dllexport) when building the DLL, __declspec(dllimport) when using the DLL, and no annotation when building or using a static library.  A #define macro is typically used to attach the appropriate annotation.  One effect of this is that the same object file can't be used in a static library and a DLL; the code has to be compiled twice with different options.

An alternative is to export a function that returns a pointer to the data, rather than exporting the data itself.  This is a common idiom, and openssl does it if **OPENSSL_EXPORT_VAR_AS_FUNCTION** is defined.  Support for this was stripped out of LibreSSL.

_Bottom line:_

I believe that using CMake EXPORT_ALL, and adding back OPENSSL_EXPORT_VAR_AS_FUNCTION, will fix the Windows DLL problem.

**Comment by kinichiro** [comment]
Hi,

Why don't you use formal Windows binary package ?
http://ftp.openbsd.org/pub/OpenBSD/LibreSSL/libressl-2.4.2-windows.zip
It contains DLLs and include files, and I think those DLLs are exported
well.

This package seems to be made by dist-win.sh in the github portable repo on
Cygwin64 with VS2013 compiler, not by using cmake and VisualStudio, though.

Best regards,

**Comment by kinichiro** [comment]
Short notice for unresolved issue of  WINDOWS_EXPORT_ ALL_SYMBOLS.
https://gitlab.kitware.com/cmake/cmake/issues/15857
This causes empty .def file.

**Comment by mcnameej** [comment]
Doh!

I confirmed the original problem with both the VS and Ninja generators, but only confirmed the fix with Ninja.  It does appear that the VS generator's implementation of WINDOWS_EXPORT_ ALL_SYMBOLS is broken in the current CMake release (broken for projects, like LibreSSL, that pull static/object libraries into shared libraries; it probably works in the simple case of shared libraries built entirely from source files).

A fix was [committed](https://gitlab.kitware.com/cmake/cmake/commit/13a6ff31becea16d567b23abc68bfa8aa75365ce) to CMake about two months ago. I assume it will be in the next release.

**Comment by kinichiro** [comment]
This is my personal thoughts,

Creating .DEF files manually is better than using WINDOWS_EXPORT_ALL_SYMBOLS,
since it will export unwanted DLLs functions, too.
With this method, we need to maintain .DEF files according to adding or removing APIs, though.

After running dist-win.sh, we can get .DEF files generated from DLLs.
We can use these 3 .DEF files as initial version.

And cmake is able to handle .DEF file in add_library() function, like this.

```
add_library(ssl-shared SHARED $<TARGET_OBJECTS:ssl-objects> ssl.def)
```

**Comment by mcnameej** [comment]
> Creating .DEF files manually is better than using WINDOWS_EXPORT_ALL_SYMBOLS,

I agree!  I didn't suggest it because it requires ongoing maintenance, and I assumed y'all wouldn't want to go down that path.

You'll still need to deal with exported data by either bringing back OPENSSL_EXPORT_VAR_AS_FUNCTION, or adding the appropriate __declspec annotations.

You can export data symbols in a DEF file, but that's only half the battle.  Exporting a data symbol really exports the address of data, and any code that uses an imported data symbol needs to know that.  The magic of __declspec(dllimport) is that the compiler adds a level of indirection to all uses of the variable (kind of like how C++ references are implemented).

----------

Without __declspec(dllimport) you end up with code like this:

Code for foo1.dll:

```C
int data1;  // "data1" is exported via DEF file.
```

Code for foo2.lib:

```C
int data2;  // "data2" is a global variable in a static library.
```

Code for app.exe:

```C
extern int* data1;  // "data1" is imported from foo1.dll.
extern int data2;  // "data2" is defined in foo2.lib.
```

Every use of ``data1`` in app.exe's code will need to be written as ``*data1``. You sometimes see this in really old Windows code.

P.S.  __declspec(dllimport) is optional for functions.  It lets the compiler optimize calls, but the code works either way.

**Comment by kinichiro** [comment]
This is my personal thoughts,

For OPENSSL_EXPORT_VAR_AS_FUNCTION, target header files are only 2, `asn1.h` and `asn1t.h` in include/openssl/.
I would like to suggest these `#ifndef OPENSSL_EXPORT_VAR_AS_FUNCTION ...` should be added by local patch in libressl-portable rather than modifing headers in OpenBSD cvs repo.

Then, #define of OPENSSL_EXPORT_VAR_AS_FUNCTION can be added in CMakeLists.txt,

```
if(MSVC)
	add_definitions(-DOPENSSL_EXPORT_VAR_AS_FUNCTION)
endif()
```

This will be also added in mingw section of m4/check-os-options.m4.

**Comment by kinichiro** [comment]
I tried to create the patch for this issue.
I added DLL export functionality to CMake, added 3 DEF files, created patches for 2 headers.

You can see diffs from here, 
https://github.com/libressl-portable/portable/compare/master...kinichiro:fix_cmake_vs
And you can download it from here,
https://github.com/kinichiro/portable/tree/fix_cmake_vs

I could generate LibreSSL.sln with this command,

```
cmake -G "Visual Studio 14 2015 Win64" -DBUILD_SHARED=on ..
```

With VS2015 Express, I was able to build exported DLLs.
After that, I wrote small libtls program and linked it with exported DLLs, and it runs OK.
(build openssl.exe fails now because of undefined PATH_MAX. This should be solved.)

I did not tested OPENSSL_EXPORT_VAR_AS_FUNCTION yet, then I can not tell it is OK or NG.

Can you see and try this patch before posting PR, please ?
You can get the tarball by running dist.sh after cloning my repo https://github.com/kinichiro/portable/tree/fix_cmake_vs .

**Comment by mcnameej** [comment]
Got your branch and built it.

I temporarily fixed the openssl build failure by adding this to apps.c and ca.c:

```C
#ifndef PATH_MAX
 #ifdef MAX_PATH
  #define PATH_MAX MAX_PATH
 #else
  #define PATH_MAX 260
 #endif
#endif
```

I also had to hack the CMake-generated openssl.vcxproj file so it linked against the DLL import libraries rather than the static libs.  Need to look at the CMakeLists.txt files.

This got me to the point where the openssl link fails with a few undefined symbols:

strcasecmp
asprintf
sleep

Unfortunately, that's all the time I have today to look at this.  Will have to pick it up again tomorrow.

**Comment by kinichiro** [comment]
I added definition of PATH_MAX in include/compat/limits.h temporary.
I posted this as another issue.

To resolve undefined symbols, I added those into crypto/crypto.def.
Those 3 functions (strcasecmp, asprintf and sleep) are existed in crypto-38.dll, but are not exported.
I am not sure that openssl.exe should be linked dynamic, though, I did this for experimental.
I thought that you just wanted to test exported DLLs linking with openssl.exe.
(IMHO, it is good openssl.exe is linked as static rather than as dynamic.)

I updated my repo https://github.com/kinichiro/portable/tree/fix_cmake_vs .
You need to tweak .vcxproj file still, but I think you can build with this update.

**Comment by kinichiro** [comment]
I took a look at how OPENSSL_EXPORT_VAR_AS_FUNCTION works.

This switches 2 sets of macros (ASN1_* and *_GLOBAL) in asn1.h, asn1t.h and e_os2.h on OpenSSL 1.0.1.
But use of these macros had removed from *.c on LibreSSL already, then I do NOT think just modifying header files solves the DLL variable export issue.
And, IMHO, we should not use these macros again in libressl-portable since it will corrupt the decisions of upstream commits.

EX1) ASN1_SEQUECE is like below. (ASN1_* has several macros, here is one of them.)

```
In OpenSSL 1.0.1 crypto/asn1/tasn_typ.c (using macro)
    116 IMPLEMENT_ASN1_TYPE(ASN1_SEQUENCE)

In LibreSSL current crypto/asn1/tsan_typ.c (no macro)
    564 const ASN1_ITEM ASN1_SEQUENCE_it = {
    565         .itype = ASN1_ITYPE_PRIMITIVE,
    566         .utype = V_ASN1_SEQUENCE,
    567         .sname = "ASN1_SEQUENCE",
    568 };
```

EX2) DES_check_key and DES_rw_mode.

```
In OpenSSL 1.0.1 (using macros)
    crypto/des/des.h:
    126 OPENSSL_DECLARE_GLOBAL(int, DES_check_key); /* defaults to false */
    127 # define DES_check_key OPENSSL_GLOBAL_REF(DES_check_key)
    128 OPENSSL_DECLARE_GLOBAL(int, DES_rw_mode); /* defaults to DES_PCBC_MODE */
    129 # define DES_rw_mode OPENSSL_GLOBAL_REF(DES_rw_mode)

    crypto/des/enc_read.c:
     68 OPENSSL_IMPLEMENT_GLOBAL(int, DES_rw_mode, DES_PCBC_MODE)

    crypto/des/set_key.c:
     70 OPENSSL_IMPLEMENT_GLOBAL(int, DES_check_key, 0)

In LibreSSL current (no macro)
    include/openssl/des.h:
    110 extern int DES_check_key;       /* defaults to false */
    111 extern int DES_rw_mode;         /* defaults to DES_PCBC_MODE */

    crypto/des/enc_read.c:
     68 int DES_rw_mode = DES_PCBC_MODE;

    crypto/des/set_key.c:
     69 int DES_check_key = 0;  /* defaults to false */
```

For EX2) DES_check_key and DES_rw_mode, I could find descriptions in `man des`.
For EX1), I have no idea in which case we need to access exported global variables in DLLs.

**Comment by mcnameej** [comment]
> use of these macros had removed from *.c on LibreSSL already ... And, IMHO, we should not use these macros again in libressl-portable since it will corrupt the decisions of upstream commits.

Didn't realize that code had been removed.  I should have checked that :angry:.  Definitely agree about not re-introducing the macros at this point.  Will have to think some more about the data problem.

In other news...

crypto.def needs to have `gettimeofday` added to it.

`gai_strerrorA` should be removed from tls.def (this is defined as an inline function in a Windows header).

I'd like to recommend three changes to the top-level CMakeLists.txt...

* Ignore warning C4267 (lots of noise, very little signal).
* Edit CMAKE_C_FLAGS, don't overwrite it.

```patch
*** kinichiro-fix_cmake_vs/CMakeLists.txt Sun Sep 11 07:47:32 2016
--- local/CMakeLists.txt Mon Sep 12 08:13:21 2016
***************
*** 105,114 ****
      "C4820" # 'bytes' bytes padding added after construct 'member_name'
      "C4996" # 'read': The POSIX name for this item is deprecated. Instead,
              # use the ISO C++ conformant name: _read.
    )
    string(REPLACE "C" " -wd" MSVC_DISABLED_WARNINGS_STR
      ${MSVC_DISABLED_WARNINGS_LIST})
!   set(CMAKE_C_FLAGS  "-MP -W4 ${MSVC_DISABLED_WARNINGS_STR}")
  endif()
  
  check_function_exists(asprintf HAVE_ASPRINTF)
--- 107,121 ----
      "C4820" # 'bytes' bytes padding added after construct 'member_name'
      "C4996" # 'read': The POSIX name for this item is deprecated. Instead,
              # use the ISO C++ conformant name: _read.
+     "C4267" # conversion from 'size_t' to 'some type that is almost
+             # certainly safe to convert a size_t to'.
    )
    string(REPLACE "C" " -wd" MSVC_DISABLED_WARNINGS_STR
      ${MSVC_DISABLED_WARNINGS_LIST})
! 
!   string(REGEX REPLACE "[/-]W[1234][ ]?" "" CMAKE_C_FLAGS ${CMAKE_C_FLAGS})
! 
!   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -MP -W4 ${MSVC_DISABLED_WARNINGS_STR}")
  endif()
  
  check_function_exists(asprintf HAVE_ASPRINTF)
```

* Allow building openssl.exe with LibreSSL DLL's.

```patch
*** kinichiro-fix_cmake_vs/CMakeLists.txt Sun Sep 11 07:47:32 2016
--- local/CMakeLists.txt Mon Sep 12 08:13:21 2016
***************
*** 223,229 ****
    endif()
  endif()
  
! set(OPENSSL_LIBS ssl crypto)
  if(CMAKE_HOST_WIN32)
    set(OPENSSL_LIBS ${OPENSSL_LIBS} ws2_32)
  endif()
--- 230,252 ----
    endif()
  endif()
  
! if(NOT (CMAKE_SYSTEM_NAME MATCHES "(Darwin|CYGWIN)"))
!   set(BUILD_SHARED true)
! endif()
! 
! # USE_SHARED builds applications (e.g. openssl) using shared LibreSSL.
! # By default, applications use LibreSSL static library to avoid dependencies.
! # USE_SHARED isn't set by default; use -DUSE_SHARED=ON with CMake to enable.
! # Can be helpful for debugging; don't use for public releases.
! if(NOT BUILD_SHARED)
!   set(USE_SHARED off)
! endif()
! 
! if(USE_SHARED)
!   set(OPENSSL_LIBS ssl-shared crypto-shared)
! else()
!   set(OPENSSL_LIBS ssl crypto)
! endif()
  if(CMAKE_HOST_WIN32)
    set(OPENSSL_LIBS ${OPENSSL_LIBS} ws2_32)
  endif()
***************
*** 238,247 ****
  endif()
  if(CMAKE_SYSTEM_NAME MATCHES "SunOS")
    set(OPENSSL_LIBS ${OPENSSL_LIBS} nsl socket)
- endif()
- 
- if(NOT (CMAKE_SYSTEM_NAME MATCHES "(Darwin|CYGWIN)"))
-   set(BUILD_SHARED true)
  endif()
  
  check_type_size(time_t SIZEOF_TIME_T)
--- 261,266 ----
```

**Comment by kinichiro** [comment]
I had done suggested, and I could verify building with VS2015 Express.
I removed 2 patches for asn1.h and asn1t.h since those would not solve anything about OPENSSL_EXPORT_VAR_AS_FUNCTION.
Now, commit logs are not organized and I should fix them after these works, before PR.
And I still would like to see for EX1) above, in which case we need to access exported global variables in DLLs.

**Comment by kinichiro** [comment]
I had re-arranged commits and created another branch fix_cmake_vs_2.
https://github.com/kinichiro/portable/tree/fix_cmake_vs_2

I will leave an original branch fix_cmake_vs for reference.
https://github.com/kinichiro/portable/tree/fix_cmake_vs

I removed PATH_MAX commit since it is not a tool chain issue.
2 commits related to OPENSSL_EXPORT_VAR_AS_FUNCTION were deleted.
All modifications are integrated into 1 commit and includes all commit logs.

I would like to post this new branch as PR "export DLLs functions for MSVC with CMake".

**Comment by busterb** [comment]
This should be working with #227
